src/hotspot/share/gc/g1/g1RemSet.cpp

```diff
@@ -1688,12 +1687,20 @@ void G1RemSet::refine_card_concurrently(CardValue* const card_ptr,
     return;
   }
 
-  // Re-dirty the card and enqueue in the *shared* queue.  Can't use
-  // the thread-local queue, because that might be the queue that is
-  // being processed by us; we could be a Java thread conscripted to
-  // perform refinement on our queue's current buffer.
+  // Re-dirty and re-enqueue the card for later processing.  We can't use
+  // the thread-local queue, because that might be the queue that is being
+  // processed by us; we could be a Java thread conscripted to perform
+  // refinement on our queue's current buffer.  This situation only arises
+  // from an extremely rare race condition in the memregion iteration, so
+  // it's not worth any significant development effort or clever lock-free
+  // queue implementation.  Instead we use brute force, allocating and
+  // enqueuing an entire buffer for just this card.
   *card_ptr = G1CardTable::dirty_card_val();
-  G1BarrierSet::shared_dirty_card_queue().enqueue(card_ptr);
+  G1DirtyCardQueueSet& dcqs = G1BarrierSet::dirty_card_queue_set();
```

Comment:
This is just a suggestion: maybe put this code to put and enqueue that card into an extra buffer into an extra method. It does not seem to fit just here.

Reply:
I split that out into new helper function enqueue_for_reprocessing.